File: code/Java/state/gumballmachine/HasQuarterState.java
package state.gumballmachine;

import java.util.Random;

public class HasQuarterState implements State{
    private GumballMachine gumballMachine;
    private Random randomWinner = new Random(System.currentTimeMillis());

    public HasQuarterState(GumballMachine gm) {
        this.gumballMachine = gm;
    }

    @Override
    public void insertQuarter() {
        System.out.println("You can't insert another quarter");
    }

    @Override
    public void ejectQuarter() {
        System.out.println("Quarter returned");
        gumballMachine.setCurState(gumballMachine.getNoQuarterState());
    }

    @Override
    public void turnCrank() {
        System.out.println("You turned...");
        int winner = randomWinner.nextInt(10);
        if((winner == 0) && (gumballMachine.getCount() > 1))
            gumballMachine.setCurState(gumballMachine.getWinnerState());
        else
            gumballMachine.setCurState(gumballMachine.getSoldState());
    }

    @Override
    public void dispense() {
        System.out.println("You need to turn the crank");
    }

    @Override
    public void refill() {    }

//    @Override
//    public String toString() {
//        return "HasQuarterState{" +
//                "gumballMachine=" + gumballMachine +
//                '}';
//    }
}

File: code/C++/state/hasquarterstate.cpp
#include "hasquarterstate.h"

HasQuarterState::HasQuarterState(GumballMachine * const gm) :gumballMachine(gm) {
	srand((int)time(NULL));
}

void HasQuarterState::insertQuarter()
{
	std::cout << "You can't insert another quarter" << std::endl;
}

void HasQuarterState::ejectQuarter()
{
	std::cout << "Quarter returned" << std::endl;
	this->gumballMachine->setState(this->gumballMachine->getNoQuarterState());
}

void HasQuarterState::turnCrank()
{
	std::cout << "You turned..." << std::endl;
	int randomNumber = rand() % 11;
	// std::cout << randomNumber << std::endl;
	if (randomNumber == 0 && this->gumballMachine->getCount() > 1)
		this->gumballMachine->setState(this->gumballMachine->getWinnerState());
	else
		this->gumballMachine->setState(this->gumballMachine->getSoldState());
}

void HasQuarterState::dispense()
{
	std::cout << "You need to turn the crank" << std::endl;
}

void HasQuarterState::refill()
{

}

void HasQuarterState::display()
{
	std::cout << "HasQuarterState, count:" << this->gumballMachine->getCount() << std::endl;
}